### PR TITLE
WIP, RFC: types (DO NOT MERGE)

### PR DESCRIPTION
**Pull request by samoojj**
Hello,

I'm opening this as a request for comments about how adding types to this library could be beneficial and helpful.

This is incomplete, and could use some more work. 

I will make specific questions in the diff.

### CODE REVIEW

File: types/test/index.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    byId: {};
+  };
+
+  const buildCat = (name: string, breed: CatBreed, birthday?: Date) =>
+    Promise.resolve(new Cat(name, breed, birthday));
+  const catStoreOptions = VuexStoreBuilder<CatState, Cat>("build", buildCat, {
+    getKey: Animal.getId
+  });
+  const catStore = new Vuex.Store(catStoreOptions);
+  catStore.dispatch("build", {
```
Comment (samoojj) [inline types/test/index.ts]: This isn't super useful as it is right now, as the actions are not enumerated.

Would it be worth having `VuexStoreBuilder(..args)` generate `{ ACTIONS, storeOptions }` where `ACTIONS` is an enum that lists the actions defined in the generated store module?

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  receivedMutationName?: string;
+  failedMutationName?: string;
+  catchBlock?: (error: any) => void;
+}
+
+export type ActionBuilder<S, T> = (
+  slug: string,
+  call: (payload?: any) => Promise<T>,
+  options: ActionBuilderOptions
+) => Action<S, any>;
```
Comment (samoojj) [inline types/index.d.ts]: `any` here is the `RootState` type. As `RootState` is derived from all non-root states.

I'm not sure how to account for `RootState`. If a store module is dynamically registered, wouldn't that mutate the type?

File: types/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "target": "es5",
+    "module": "es2015",
+    "moduleResolution": "node",
+    "lib": ["es5", "dom", "es2015.promise"],
+    "strict": true,
+    "noEmit": true
+  },
+  "include": ["*.d.ts", "../dist/logger.d.ts"]
```
Comment (samoojj) [inline types/tsconfig.json]: Logger? I don't remember what that is...

File: types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  receive?: MutationMethod;
+  fail?: MutationMethod;
+  state?: S;
+  getters?: GetterTree<S, any>;
+  mutations?: MutationTree<S>;
+  actions?: ActionTree<S, any>;
+  modules?: ModuleTree<S>;
+}
+
+export declare function VuexStoreBuilder<S extends { byId: Dictionary<T> }, T>(
```
Comment (samoojj) [inline types/index.d.ts]: Is requiring `S extends { byId: Dictionary<T> }` a good idea?

In some ways, this requirement is not really a requirement, as it's possible to override `state` and change its shape.

File: types/test/index.ts (hunk 1 above)
Comment (samoojj) [reply types/test/index.ts]: Is that even possible?

File: src/actions.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  call: (payload?: any) => Promise<T>,
+  {
+    requestedMutationName = requested(slug),
+    receivedMutationName = received(slug),
+    failedMutationName = failed(slug),
+    catchBlock = function(error) {
+      this.commit(failedMutationName, error);
+    }
+  }: ActionBuilderOptions = {}
+): ActionHandler<S, R> => async ({ commit }, payload) => {
```
Comment (samoojj) [inline src/actions.ts]: `R` should be `any`

File: types/test/index.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    name: "Oliver",
+    breed: CatBreed.SNOWSHOE_SIAMESE,
+    birthday: new Date("06/01/2010")
+  });
+  catStore.dispatch("build", {
+    name: "Max",
+    breed: CatBreed.MAINE_COON,
+    birthday: new Date("02/25/2010")
+  });
```
Comment (kristinruben) [inline types/test/index.ts]: 😻

File: src/actions.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import { requested, received, failed } from "./strings";
+import { ActionHandler } from "vuex/types";
+import { ActionBuilderOptions } from "../types";
+
+export const defaultActionBuilder = <S, R, T>(
```
Comment (samoojj) [inline src/actions.ts]: `R` should be removed

**Comment by samoojj** [comment]
I'm having some trouble getting the actual `vuexStoreBuilder` function typed.
I have a type for it, `VuexStoreBuilder<S extends { byId?: Dictionary<T> }, T>`.
But I haven't been able to actually create the function without resolving the generics. I'm not sure how to proceed, and would appreciate any input about that.

File: types/index.d.ts (hunk 2 above)
Comment (maxbogue) [reply types/index.d.ts]: Why does it need the `RootState` and not only the state of the module?

Comment (samoojj) [reply types/index.d.ts]: @maxbogue I think that the `Action` type provided by `vuex/types` needs the RootState. 
https://github.com/vuejs/vuex/blob/dev/types/index.d.ts#L107

Perhaps I'm misunderstanding the generics?

**Pull request closed without merging**